### ISSUE TITLE
[Feature request] Throw an error if configuration is insufficient

### ISSUE THREAD

**Issue by northivanastan** [robiso/wondercms#73]
If a PHP library is missing or permissions are incorrect, WonderCMS (in my testing) simply either shows a blank page or redirects improperly.

I would prefer that WonderCMS handle these situations by displaying an error.

**Comment by robiso** [comment]
I would also love to see this implemented. Would you be possibly interested in making a pull request?

**Comment by robiso** [comment]
Here's an idea if anyone want's to run with it. A couple of functions that check for the minimum WonderCMS requirements.

It checks if the three extensions are installed on the server (cURL, Zip, mbstring) and additionally if the PHP version is above 5.5. 

```
    private static function isCurlInstalled() {
        if  (in_array  ('curl', get_loaded_extensions())) {
            return true;
        }
        else {
            return false;
        }
    }

    private static function isMbstringInstalled() {
        if  (in_array  ('mbstring', get_loaded_extensions())) {
            return true;
        }
        else {
            return false;
        }
    }

    private static function isZipInstalled() {
        if  (in_array  ('zip', get_loaded_extensions())) {
            return true;
        }
        else {
            return false;
        }
    }

    private static function isPHPversionOK() {
        if (version_compare(phpversion(), "5.5.0", ">=")) {
            return true;
        }
        else {
            return false;
        }
    }
```

The above functions would have to be called in the init() function and additionally run before the createDatabase() function.

I'm sure better solutions can come out of this, if there's enough interest, it will be integrated into WonderCMS.

**Comment by NicolasCARPi** [comment]
@robiso instead of looping on the array everytime, use [extension_loaded](https://secure.php.net/manual/en/function.extension-loaded.php) ;)

And instead of returning bool, throw a custom exception.

BTW, this code can be written like this:

~~~php
return \extension_loaded('zip');
~~~

And you can skip the `else` part:

~~~php
if (version_compare(phpversion(), "5.5.0", ">=")) {
            return true;
        }
        return false;
}
~~~

**Comment by robiso** [comment]
Much cleaner and neater @NicolasCARPi. Thank you!

Is there any clean way this can be wrapped in one function with returning error messages on what requirement is not met like "Missing mod_rewrite" or "Insufficient PHP version"?

If you share your first/last name and website/Twitter handle, I'll gladly add you to the WonderCMS contributors list: https://wondercms.com/special-contributors and in the release notes if this gets included.

**Comment by NicolasCARPi** [comment]
One way to do it:

~~~php
$extensions = array('curl', 'zip', 'mbstring');
    foreach ($extensions as $ext) {
        if (!extension_loaded($ext)) {
            $message = "The " . $ext . " extension is NOT loaded.
                    <a href='https://doc.example.com'>Click here to read how to fix this.</a>";
           throw new MissingExtensionException($message);
        }
    }
~~~

But I'm not familiar with your codebase so do as you like.


--------------------------------------

My advice is to try and start a new project (something completely different) with Symfony (or any other PHP framework) , and get a feeling of what is modern PHP. Then you'll come back to this project and realize it's not really viable. With opcache and templates and other optimizations, I think the difference in speed (because that's what you're aiming with a single file app right?) is negligeable.

Also, if you have a complete normal PHP project, you can let composer deal with PHP min version and extensions (and remove completely any code related to this). Because the `composer install` step will do that for you.

I'd love to contribute to your project but (and please take this as constructive criticism), you're too far behind! I believe what you are lacking is knowledge about how it's done by others, hence my suggestion of trying a new project in a framework. You have to get out of your confort zone or you'll stagnate. Also, drop version < 7.1, they are not maintenaid anymore and one should not code against it but aim for newer version (with new features like type hinting and new operators).

Once you start to understand how the different pieces can play together, you'll never want to put HTML in a PHP file again ;)

There is no shame in using libraries that other talented devs wrote. For instance, instead of using the $_POST and $_SESSION, you can use symfony/http-foundation to get a nice OO interface and other cool stuff.

I'll stop here because there are lot of things to say! Cheers,
~Nico

**Comment by robiso** [comment]
@NicolasCARPi thank you very much! Your criticism is encouraged, accepted and appreciated. :)

While being a mediocre developer myself, I've always found it challenging to do a lot of the stuff you've mentioned. That's also one of the few reasons why I haven't pursued a strictly developer career.

I have a couple questions regarding a development of a new CMS/app:
- How would users on shared hosting packages install the new CMS (if they can't use composer install?)
  - Is there a possibility of them simply unzipping and uploading files to their server?
- How would a web service like Microsoft Azure ([which currently offers WonderCMS as a quick CMS as the only option against WordPress)](https://azure.microsoft.com/en-gb/try/app-service/web/?language=php) or Softaculous [(which enables installing WonderCMS through cPanel)](https://www.softaculous.com/apps/cms/WonderCMS) deal with a small framework?

**Comment by NicolasCARPi** [comment]
> How would users on shared hosting packages install the new CMS (if they can't use composer install?)

Why wouldn't they be able to run composer install? Composer doesn't need any root permissions and can be run on shared hosting. But nowadays I think shared hosting is a thing of the past, as you can have a VPS (where you are root) for a very cheap price (try [this referral link](https://m.do.co/c/c2ce8f861e0e) and get $10 to try it out!). My go-to solution would be to use a docker image. So users would only need to run "docker run -p 80:80 robiso/wondercms" and that's it (in its most simple form). And because the docker image would provide all the extensions, you're sure everything will run smoothly. Of course, users would need docker, but on DigitalOcean you can have a server with Docker in a few seconds so…

> Is there a possibility of them simply unzipping and uploading files to their server?

I think that's the old school way of doing things. Nowadays a "build step" is preferred. This build step will install dependencies and minify assets (CSS, JS) for instance. You could also provide on the release page a full zip file with everything packed inside, but then you cannot be sure the user will be able to run it, hence the docker solution where you are master of everything.

I don't know about Azure, but I know Softaculous was working fine with a composer install step. It's really something VERY standard in all php projects. In fact a lot of environment will detect a composer.json and run composer install by themselves (think CI envs).

> You'll be on the same list as the CSS inventor. :)

Ahah, ok then you convinced me :p My name is my username and my twitter is @elabftw.

As a side note: I see you are providing .htaccess files. This is not great for a project aiming at speed. See [this link](https://stackoverflow.com/questions/25063677/how-much-does-using-htaccess-files-slow-down-website-performance-especially-wit). Also, what if I use nginx or caddy? Again, we endup looking at docker where everything is configured and working fine because YOU configured it before shipping to users.

If I find the time and the will, I'll try and make a PR with a composer.json and Dockerfile so you can see more clearly what I'm talking about. The idea is to provide a docker image AND also allow it to run without docker (if the environment is complete).

Cheers,
~Nico

**Comment by robiso** [comment]
@NicolasCARPi, @northivanastan, here's a function for the future 3.0.0 version for checking the minimum requirements.
```php
private static function checkMinimumRequirements()
{
    if (version_compare(phpversion(), "7.1", "<=")) {
        echo "<p>To run WonderCMS, PHP version 7.1 or greater is required.</p>";
    }
    $extensions = array('curl', 'zip', 'mbstring');
    foreach ($extensions as $ext) {
        if (!extension_loaded($ext)) {
            $message = "<p>The <b>" . $ext . "</b> extension is missing. Please contact your host or configure your server to have the <b>" . $ext . "</b> extension enabled.</p>";
            echo $message;
        }
    }
}
```
I'm hoping this will suffice.

**Comment by NicolasCARPi** [comment]
Well, you don't have custom exceptions (yet) but instead of echoing the message, maybe use:

~~~php
throw new RuntimeException($message);
~~~

so the code execution stops there.

And the code calling it would catch the Exception, and display the error to the user.

Right now, the function displays a message but gives no information if the requirements are OK. You could also add a `return false` after the `echo` lines, and `return true` at the end, so when you call it you can do:

~~~php
if (wCMS::checkMininumRequirements() !== true) {
die or something;
}
~~~

**Comment by robiso** [comment]
@NicolasCARPi, with the latter implemented in the following way
```php
private static function checkMinimumRequirements()
{
    if (version_compare(phpversion(), "7.1", "<=")) {
        $message = "<p>To run WonderCMS, PHP version 7.1 or greater is required.</p>";
    }
    $extensions = array('curl', 'zip', 'mbstring');
    foreach (!$extensions as $ext) {
        if (extension_loaded($ext)) {
            $message .= "<p>The <b>" . $ext . "</b> extension is missing. Please contact your host or configure your server to have the <b>" . $ext . "</b> extension enabled.</p>";
            echo $message;
            return false;
        }
        return true;
    }
}
```
and adding
```php
if (wCMS::checkMinimumRequirements() !== true) {
    die;
}
```
to the createDatabase function, I only receive the following two error messages, instead all of them:

> To run WonderCMS, PHP version 7.1 or greater is required.
> The curl extension is missing. Please contact your host or configure your server to have the curl extension enabled.

The last two error messages (zip and mbstring) are missing. I tried switching around my return true's and false's and couldn't yield and positive results that would include all of the necessary error messages.

**Comment by NicolasCARPi** [comment]
That's because the return true is inside the foreach loop ;)

**Comment by robiso** [comment]
@NicolasCARPi, I was hoping that's the only thing I missed, unfortunately I've already tried this.
So even if I do:
```php
private static function checkMinimumRequirements()
{
    if (version_compare(phpversion(), "7.1", "<=")) {
        $message = "<p>To run WonderCMS, PHP version 7.1 or greater is required.</p>";
    }
    $extensions = array('curl', 'zip', 'mbstring');
    foreach ($extensions as $ext) {
        if (!extension_loaded($ext)) {
            $message .= "<p>The <b>" . $ext . "</b> extension is missing. Please contact your host or configure your server to have the <b>" . $ext . "</b> extension enabled.</p>";
            echo $message;
            return false;
        }
    }
    return true;
}
```
I turn into the one foreach errrr message once (expected the zip and mbstring)
> To run WonderCMS, PHP version 7.1 or greater is required.
> The curl extension is missing. Please contact your host or configure your server to have the curl extension enabled.

I'm positive is something else silly I've done.

**Comment by NicolasCARPi** [comment]
Yes, you run into the error message only once. As expected (as the function returns). If you wish to show all the extensions missing at once (which is indeed a good thing to do), maybe set a flag like so:

~~~php
private static function checkMinimumRequirements()
{
    if (version_compare(phpversion(), "7.1", "<=")) {
        $message = "<p>To run WonderCMS, PHP version 7.1 or greater is required.</p>";
    }
    $extensions = array('curl', 'zip', 'mbstring');
    $missingExtensions = array();
    foreach ($extensions as $ext) {
        if (!extension_loaded($ext)) {
            $missingExtensions[] = $ext;
       }
    }
    if (!empty($missingExtensions)) {
            $message .= "<p>The following extensions are required: " . implode($missingExtensions, ', ') . ". Please contact your host or configure your server to enable them.</p>";
            echo $message;
            return false;
    }
    return true;
}
~~~

Of course you might want to count the extensions for the plural form of the sentence ;)

**Comment by robiso** [comment]
@NicolasCARPi, it was completely unclear to me how the previous function returned all of the required error messages at once and the new function didn't. Thank you very much for the awesome explanation and fix, appreciated. 😃

**Comment by robiso** [comment]
Implemented in experimental 3.0.0, which is looking to become a new stable release soon.
Thanks for helping out!
